vsmartcard-vpcd, vsmartcard-pcsc-relay: init at 0.9-unstable-2025-01-25, nixos/vsmartcard-vpcd: init

StarGate01 opened the pull request:
###### Description of changes

This PR adds the virtual smart card reader and relay: https://frankmorgner.github.io/vsmartcard/ . It can be used to e.g. emulate a smartcard when combined with e.g. https://jcardsim.org/ , or to debug smartcard communications via an Android phone bridge.

The service module `service.services.vsmartcard-vpcd` is used to enable and configure the virtual smart card driver. It appends a configuration to the `pcscd` `readers.conf`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

StarGate01 committed:
vsmartcard-vpcd: init at 0.9-unstable-2025-01-25

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -14443,6 +14443,14 @@ with pkgs;
     SDL = SDL_sixel;
   };
 
+  vsmartcard-vpcd = callPackage ../by-name/vs/vsmartcard-vpcd/package.nix {
+    inherit (darwin.apple_sdk.frameworks) PCSC;
+  };
+
+  vsmartcard-pcsc-relay = callPackage ../by-name/vs/vsmartcard-pcsc-relay/package.nix {
+    inherit (darwin.apple_sdk.frameworks) PCSC;
+  };
+
```
We can avoid this entry by using darwin as an input which is preferable anyway

SuperSandro2000 commented on pkgs/by-name/vs/vsmartcard-vpcd/package.nix:
```diff
@@ -0,0 +1,54 @@
+    ]))
+  ] ++ lib.optionals stdenv.isDarwin [ PCSC ];
+
+  configureFlags = [ ] ++ lib.optional stdenv.isDarwin "--enable-infoplist";
```
```suggestion
  configureFlags = lib.optional stdenv.isDarwin "--enable-infoplist";
```
lib.optional returns an empty list when the condition is false

SuperSandro2000 commented on pkgs/by-name/vs/vsmartcard-vpcd/package.nix:
```diff
@@ -0,0 +1,54 @@
+      pillow
+      gnureadline
+    ]))
+  ] ++ lib.optionals stdenv.isDarwin [ PCSC ];
```
```suggestion
  ] ++ lib.optionals stdenv.isDarwin [ darwin.apple_sdk.frameworks.PCSC ];
```

SuperSandro2000 commented on pkgs/by-name/vs/vsmartcard-vpcd/package.nix:
```diff
@@ -0,0 +1,54 @@
+    sha256 = "sha256-o01ASvHa68++wrHPLYxd452kHM0EAYQuMX8sbGjvvsg=";
+  };
+
+  sourceRoot = "source/virtualsmartcard";
```
```suggestion
  sourceRoot = "${src.name}/virtualsmartcard";
```

SuperSandro2000 commented on pkgs/by-name/vs/vsmartcard-vpcd/package.nix:
```diff
@@ -0,0 +1,54 @@
+, pkg-config
+, libtool
+, autoreconfHook
+, pcsclite, PCSC
```
```suggestion
, pcsclite
, darwin
```

SuperSandro2000 commented on pkgs/by-name/vs/vsmartcard-vpcd/package.nix:
```diff
@@ -0,0 +1,54 @@
+
+stdenv.mkDerivation rec {
+  pname = "vsmartcard-vpcd";
+  version = "20220814";
```
```suggestion
  version = "0.9-unstable-2022-08-14";
```

SuperSandro2000 commented on pkgs/by-name/vs/vsmartcard-pcsc-relay/package.nix:
```diff
@@ -0,0 +1,54 @@
+  meta = with lib; {
+    description = "Relays a smart card using an contact-less interface";
+    homepage = "https://frankmorgner.github.io/vsmartcard/pcsc-relay/README.html";
+    license = licenses.gpl3;
```
We should use gpl3Only or gpl3Plus

SuperSandro2000 commented on pkgs/by-name/vs/vsmartcard-pcsc-relay/package.nix:
```diff
@@ -0,0 +1,54 @@
+      pillow
+      gnureadline
+    ]))
+  ] ++ lib.optionals stdenv.isDarwin [ PCSC ];
```
```suggestion
  ] ++ lib.optionals stdenv.isDarwin [ darwin.apple_sdk.frameworks.PCSC ];
```

SuperSandro2000 commented on pkgs/by-name/vs/vsmartcard-pcsc-relay/package.nix:
```diff
@@ -0,0 +1,54 @@
+    sha256 = "sha256-o01ASvHa68++wrHPLYxd452kHM0EAYQuMX8sbGjvvsg=";
+  };
+
+  sourceRoot = "source/pcsc-relay";
```
```suggestion
  sourceRoot = "${src.name}/pcsc-relay";
```

SuperSandro2000 commented on pkgs/by-name/vs/vsmartcard-pcsc-relay/package.nix:
```diff
@@ -0,0 +1,54 @@
+
+stdenv.mkDerivation rec {
+  pname = "vsmartcard-pcsc-relay";
+  version = "20220814";
```
```suggestion
  version = "0.9-unstable-2022-08-14";
```

SuperSandro2000 commented on pkgs/by-name/vs/vsmartcard-pcsc-relay/package.nix:
```diff
@@ -0,0 +1,54 @@
+, gengetopt
+}:
+
+stdenv.mkDerivation rec {
```
Could we inherit version and src from vsmartcard-vpcd?

StarGate01 committed:
nixos/vsmartcard-vcpd: init

StarGate01 committed:
vsmartcard-pcsc-relay: init at 0.9-unstable-2025-01-25

StarGate01 commented:
@SuperSandro2000 I have implemented all of your suggestions, thank you!

I have also added a service module definition to make configuring the PCSC driver just a simple matter of 
```
services.vsmartcard-vpcd = {
  enable = true;
};
```

The `pcscd` option `readerConfig` was upgraded to a list `readerConfigs`, to allow collection of config fragment from multiple sources. The old option was marked as deprecated but continues to work.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4220

SuperSandro2000 commented on nixos/modules/services/hardware/pcscd.nix:
```diff
@@ -17,6 +17,15 @@ let
 in
 {
+  imports = [
+    (lib.mkChangedOptionModule [ "services" "pcscd" "readerConfig"] [ "services" "pcscd" "readerConfigs" ]
+     (config:
+        let readerConfig = getAttrFromPath [ "services" "pcscd" "readerConfig"] config;
```
```suggestion
    (lib.mkChangedOptionModule [ "services" "pcscd" "readerConfig" ] [ "services" "pcscd" "readerConfigs" ]
     (config:
        let readerConfig = getAttrFromPath [ "services" "pcscd" "readerConfig" ] config;
```

SuperSandro2000 commented on nixos/modules/services/development/vsmartcard-vpcd.nix:
```diff
@@ -0,0 +1,41 @@
+      FRIENDLYNAME "Virtual PCD"
+      DEVICENAME   ${cfg.hostname}:0x${lib.toHexString cfg.port}
+      LIBPATH      ${pkgs.vsmartcard-vpcd}/var/lib/pcsc/drivers/serial/libifdvpcd.so
+      CHANNELID    0x${lib.toHexString cfg.port}''
+    ];
```
```suggestion
      CHANNELID    0x${lib.toHexString cfg.port}
    '' ];
```

SuperSandro2000 commented on nixos/modules/services/development/vsmartcard-vpcd.nix:
```diff
@@ -0,0 +1,41 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
```
```suggestion
```
we should avoid adding new ones like that. Just prefix the relevant functions with lib.

StarGate01 commented:
Thank you @SuperSandro2000 for you suggestions, I have implemented them.

StarGate01 commented:
@SuperSandro2000 I have rebased and updated this PR, brought it up to modern standards and added a release notes entry. Are there any changes you still suggest?

StarGate01 commented:
Since #379118 has been merged, the extensible pcsclite serial reader configuration mechanism is available. This is used in this new module as well to enable the vsmartcard reader.

misuzu commented on nixos/modules/services/development/vsmartcard-vpcd.nix:
```diff
@@ -0,0 +1,47 @@
+
+in
+{
+
```
```suggestion
  meta.maintainers = with lib.maintainers; [ stargate01 ];
```

misuzu commented on pkgs/by-name/vs/vsmartcard-vpcd/package.nix:
```diff
@@ -0,0 +1,62 @@
+    description = "Emulates a smart card and makes it accessible through PC/SC";
+    homepage = "http://frankmorgner.github.io/vsmartcard/virtualsmartcard/README.html";
+    license = lib.licenses.gpl3Only;
+    platforms = lib.platforms.all;
```
Doesn't build on macOS:
```
  CCLD     libifdvpcd.la
Undefined symbols for architecture arm64:
  "_log_msg", referenced from:
      _IFDHCreateChannel in libifdvpcd_la-ifd-vpcd.o
      _IFDHCreateChannelByName in libifdvpcd_la-ifd-vpcd.o
      _IFDHControl in libifdvpcd_la-ifd-vpcd.o
      _IFDHCloseChannel in libifdvpcd_la-ifd-vpcd.o
      _IFDHGetCapabilities in libifdvpcd_la-ifd-vpcd.o
      _IFDHSetCapabilities in libifdvpcd_la-ifd-vpcd.o
      _IFDHSetProtocolParameters in libifdvpcd_la-ifd-vpcd.o
      ...
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

```suggestion
    platforms = lib.platforms.all;
    broken = stdenv.isDarwin;
```

StarGate01 replied:
Okay, I'll mark Darwin as broken for now. I  considered setting platforms to Linux only, since I don't have a Darwin system to test. But the upstream supports it so I'll leave it for interested contributors to jump off from.
